### PR TITLE
Improve the screen reader experience for the header menu

### LINKED ISSUE #1340: Review if `aria-label` in header navigation is necessary
The [header with navigation](https://design-system.service.gov.uk/components/header/with-service-name-and-navigation/index.html) throws the following warning in https://validator.w3.org/:
> Possible misuse of aria-label.

This relates to https://github.com/alphagov/govuk-frontend/blob/fea29aaa833d5f8233c71e0d076923a697e752cd/src/components/header/template.njk#L68 which was inherited from [govuk_template](https://alphagov.github.io/govuk_template/example-proposition-menu.html).

The [WCAG specs ](https://www.w3.org/TR/WCAG20-TECHS/ARIA6.html#ARIA6-examples) state that `aria-label` can be used for:
- ` Distinguishing navigation landmarks`
- `Identifying region landmarks`

We should review whether having `aria-label` in this context provides useful information for assistive technologies. 

There is also a related issue where we hide the list inside the navigation element when showing/hiding the navigation. This means that a user can end up with an empty `<nav></nav>` element, which may cause issues/confusion for screenreader users. See [comment](https://github.com/alphagov/govuk-frontend/issues/1340#issuecomment-501318558) for more context.

## Done when

- [ ] TBC
- [ ] "Details of breaking change" section below completed to help with writing release notes

## Details of breaking change

- which users are affected: TBC
- the change that’s been made: TBC
- changes users will have to make: TBC
- impact of users not making those changes: TBC

### PR DESCRIPTION
**Pull request by 36degrees**
This brings the use of `aria-label` and `aria-controls` in the menu navigation in the header in line with other 'best practise' examples and guidance from W3C.

## Move the label from the `<ul>` to the `<nav>`

Move the 'Navigation menu' label from the `<ul>` to the parent `<nav>` element - this has [previously been flagged as a 'possible misuse of `aria-label`'](https://github.com/alphagov/govuk-frontend/issues/1340#issue-443492815).

According to [WAI ARIA Practises 1.2][2], it's recommended to label navigation regions, especially as ['if a page includes more than one navigation landmark, each should have a unique label'][3].

Lists can optionally be labelled, but this is only '_potentially_ beneficial for users of screen readers that support both list names and navigation among lists on a page'.

Applying the label to both the list and the parent navigation could be noisy – "Potentially a source of distracting or undesirable screen reader verbosity, especially if nested within a named container, such as a navigation region."

The [provided examples][4] also apply the label consistently to the navigation region (albeit using `aria-labelledby`).

## Remove redundant 'navigation' from the nav label

In line with the section ['Composing Effective and User-friendly Accessible Names' in WAI-ARIA Authoring Practices 1.2][5], remove the word navigation from the aria-label on the navigation.

> 'Do NOT include a WAI-ARIA role name in the accessible name. For example, do not include [..] the word "navigation" in the name of a navigation region. Doing so would create duplicate screen reader output since screen readers convey the role of an element in addition to its name.'

Without this change, assistive technologies report this as 'Navigation menu navigation'.

Also remove the word 'navigation' from the button that shows or hides the menu, as it also seems redundant – the word 'menu' should be descriptive enough for users to understand what it represents, especially when encountered inside a header landmark.

## Move the button inside the `<nav>`

Move the menu button inside the navigation region, so that it is discoverable by screen reader users even when the nav is closed:

> Place mobile open/close buttons within the `<nav>` element and use them to toggle state of another child wrapper of the nav. This will ensure that the navigation landmark is still discoverable by screen readers, even if it is in a closed/hidden state.

https://a11y-style-guide.com/style-guide/section-navigation.html#kssref-navigation-navigation-mobile

> Where a lot of people go wrong is by placing the button outside the region. This would mean screen reader users who move to the `<nav>` using a shortcut would find it to be empty, which isn’t very helpful.

https://inclusive-components.design/menus-menu-buttons/#placement

> For instance, a navigation on small screen devices may initially contain only a single toggle button. But that toggle button would reveal additional links and other navigational elements, when activated.

https://www.scottohara.me/blog/2018/03/03/landmarks.html#navigation

[1]: #1340 (comment)
[2]: https://www.w3.org/TR/wai-aria-practices-1.1/#naming_role_guidance
[3]: https://www.w3.org/TR/wai-aria-practices-1.1/#aria_lh_navigation
[4]: https://www.w3.org/TR/wai-aria-practices-1.1/examples/landmarks/navigation.html
[5]: https://www.w3.org/TR/wai-aria-practices-1.2/#naming_effectively

### Before
```html
<button type="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="Show or hide navigation menu" aria-expanded="false">Menu</button>
<nav>
  <ul id="navigation" class="govuk-header__navigation " aria-label="Navigation menu">
      <li class="govuk-header__navigation-item govuk-header__navigation-item--active">
        <a class="govuk-header__link" href="#1">Navigation item 1</a>
      </li>
      <li class="govuk-header__navigation-item">
        <a class="govuk-header__link" href="#2">Navigation item 2</a>
      </li>
      <li class="govuk-header__navigation-item">
        <a class="govuk-header__link" href="#3">Navigation item 3</a>
      </li>
      <li class="govuk-header__navigation-item">
        <a class="govuk-header__link" href="#4">Navigation item 4</a>
      </li>
  </ul>
</nav>
```

### After

```html
<nav aria-label="Menu">
  <button type="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="Show or hide menu">Menu</button>

  <ul id="navigation" class="govuk-header__navigation " >
    <li class="govuk-header__navigation-item govuk-header__navigation-item--active">
      <a class="govuk-header__link" href="#1">Navigation item 1</a>
    </li>
    <li class="govuk-header__navigation-item">
      <a class="govuk-header__link" href="#2">Navigation item 2</a>
    </li>
    <li class="govuk-header__navigation-item">
      <a class="govuk-header__link" href="#3">Navigation item 3</a>
    </li>
    <li class="govuk-header__navigation-item">
      <a class="govuk-header__link" href="#4">Navigation item 4</a>
    </li>
  </ul>
</nav>
```

### JAWS / Chrome

| Before                                                                       | After                                                                                   |                                                                                          |
|------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------|
| Behaviour when focusing menu button (mobile viewport)                        | Show or hide navigation menu button collapsed, to activate press enter                  | Show or hide menu button collapsed, to activate press enter                              |
| Behaviour when entering navigation region (by tabbing to first link)         | Navigation region, navigation menu list with 8 items. Departments same page link.       | Menu navigation region, list with 8 items. Departments same page link.                   |
| Behaviour when jumping to controlled element (Insert + Alt + M)              | Move to controlled element, list, tab [visible focus moves to the list inside the menu] | Move to controlled element, show or hide menu [visible focus remains on the menu button] |
| Appearance in list of navigation landmarks (Ctrl + Insert + R)               | Banner -> Navigation                                                                    | Banner -> Menu Navigation                                                                |
| Appearance in list of navigation landmarks with nav closed (mobile viewport) | Does not appear in list of landmarks                                                    | Does not appear in list of landmarks                                                     |
| Appearance in list of navigation landmarks with nav open (mobile viewport)   | Banner -> Navigation                                                                    | Banner -> Menu Navigation                                                                |

### NVDA / Firefox

|                                                                              | Before                                                                 | After                                                        |
|------------------------------------------------------------------------------|------------------------------------------------------------------------|--------------------------------------------------------------|
| Behaviour when focusing menu button (mobile viewport)                        | Show or hide navigation menu button collapsed                          | Menu navigation landmark, show or hide menu button collapsed |
| Behaviour when entering navigation region (by tabbing to first link)         | navigation landmark Navigation menu list with 8 items Departments link | list with 8 items Departments link                           |
| Appearance in list of navigation landmarks (Insert + F7)                     | Banner -> Navigation                                                   | Banner -> Menu; navigation                                   |
| Appearance in list of navigation landmarks with nav closed (mobile viewport) | Does not appear in list of landmarks                                   | Banner -> Menu; navigation                                   |
| Appearance in list of navigation landmarks with nav open (mobile viewport)   | Banner -> Navigation                                                   | Banner -> Menu; navigation                                   |

### VoiceOver / Safari (macOS)

|                                                                         | Before                                          | After                                                  |
|-------------------------------------------------------------------------|-------------------------------------------------|--------------------------------------------------------|
| Behaviour when focusing menu button (mobile viewport)                   | Show or hide navigation menu, collapsed, button | Show or hide menu, collapsed, button, Menu, navigation |
| Behaviour when entering navigation region (by tabbing to first link)    | link, Departments, list Navigation menu 8 items | link, Departments, list 8 items                        |
| Appearance in list of landmarks rotor (VO + U)                          | Navigation                                      | Menu navigation                                        |
| Appearance in list of landmarks rotor with nav closed (mobile viewport) | Does not appear                                 | Menu navigation                                        |
| Appearance in list of landmarks rotor with nav open (mobile viewport)   | Navigation                                      | Menu navigation                                        |

### VoiceOver / Safari (iOS)

|                                                               | Before                                                                                      | After                                                                                  |
|---------------------------------------------------------------|---------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------|
| Swiping to menu button                                        | Show or hide navigation menu, button, end banner, collapsed. Double tap to expand.          | Show or hide menu, button, menu, navigation landmark, collapsed. Double tap to expand. |
| Swiping to first link in navigation region                    | Departments, link, list start, navigation menu, landmark                                    | Departments, link, list start                                                          |
| Swiping by landmark (rotor set to landmarks, swiping up/down) | Departments, link, list start, navigation menu, landmark                                    | Departments, link, list start, landmark                                                |
| Swiping by landmark with nav closed                           | Swipes between start of header (GOV.UK logo) and main region ('Press release'), missing nav | Show or hide menu, button, menu, navigation landmark, collapsed. Double tap to expand. |

### TalkBack (Android)

| | Before | After |
|---|---|---|
| Swiping to menu button | Collapsed, "Show or hide navigation menu", button. Double tap to activate, actions available, use tap with three fingers to view. | Collapsed, "Show or hide menu", button. Double tap to activate, actions available, use tap with three fingers to view. |
| Swiping to first link in navigation region | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. |
| Swiping by landmark with menu collapsed (reading controls set to 'landmarks', swiping up/down) | Navigation. | "menu", navigation. |
| Swiping by landmark with menu expanded (reading controls set to 'landmarks', swiping up/down) | Navigation, "navigation menu", list. | "menu", navigation, list. |

Tested by @lfdebrux in https://github.com/alphagov/govuk-frontend/pull/2427#issuecomment-979983062

Closes #1340

### CODE REVIEW

**Comment by 36degrees** [comment]
Raised as a draft for now – still need to do testing and write up an entry for the release notes.

**Comment by 36degrees** [comment]
Still needs testing with Android / TalkBack.

**Comment by lfdebrux** [comment]
Tested Talkback / Chrome (Android) with a Google Pixel 3 running Android 12

Before you cannot navigate to the menu navigation using the landmarks control unless the menu is already open.

Note also that by default Talkback doesn't give you the option to navigate by landmarks, the user has to enable it in the Talkback settings (see [note on Deque university page starting "some ... reading controls are hidden by default"](https://dequeuniversity.com/screenreaders/talkback-shortcuts)).

| | Before | After |
|---|---|---|
| Swiping to menu button | Collapsed, "Show or hide navigation menu", button. Double tap to activate, actions available, use tap with three fingers to view. | Collapsed, "Show or hide menu", button. Double tap to activate, actions available, use tap with three fingers to view. |
| Swiping to first link in navigation region | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. |
| Swiping by controls (reading controls set to 'controls', swiping up/down) | Collapsed, "Show or hide navigation menu", button. Double tap to activate, actions available, use tap with three fingers to view. | |
| Swiping by landmark with menu collapsed (reading controls set to 'landmarks', swiping up/down) | Navigation. | Does not appear in list of landmarks |
| Swiping by landmark with menu expanded (reading controls set to 'landmarks', swiping up/down) | Navigation, "navigation menu", list. | "menu", navigation, list. |

**Comment by lfdebrux** [comment]
Testing commit [Move button inside navigation region](https://github.com/alphagov/govuk-frontend/pull/2427/commits/537651db9ddd995b9fa440381a1d483bc25e95eb) with Talkback / Chrome (Android)

| | Before | After |
|---|---|---|
| Swiping to menu button | Collapsed, "Show or hide navigation menu", button. Double tap to activate, actions available, use tap with three fingers to view. | Collapsed, "Show or hide menu", button. Double tap to activate, actions available, use tap with three fingers to view. |
| Swiping to first link in navigation region | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. | "Departments", link. Double tap to activate, links available, use tap with three fingers to view. |
| Swiping by landmark with menu collapsed (reading controls set to 'landmarks', swiping up/down) | Navigation. | "menu", navigation. |
| Swiping by landmark with menu expanded (reading controls set to 'landmarks', swiping up/down) | Navigation, "navigation menu", list. | "menu", navigation, list. |

File: src/govuk/components/header/template.njk [hunk 1]
```diff
@@ -54,9 +54,10 @@
     </a>
     {% endif %}
     {% if params.navigation %}
-    <button type="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="{{ params.menuButtonLabel | default('Show or hide navigation menu') }}">Menu</button>
-    <nav>
-      <ul id="navigation" class="govuk-header__navigation {{ params.navigationClasses if params.navigationClasses }}" aria-label="{{ params.navigationLabel | default('Navigation menu') }}">
+    <nav aria-label="{{ params.navigationLabel | default('Menu') }}">
+      <button type="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="{{ params.menuButtonLabel | default('Show or hide menu') }}">Menu</button>
+
+      <ul id="navigation" class="govuk-header__navigation {{ params.navigationClasses if params.navigationClasses }}" >
```
Comment (36degrees) [inline src/govuk/components/header/template.njk:60]: I've left the `__navigation` element and any passed `navigationClasses` on the `<ul>`. There's an argument for moving them to the `<nav>` as well, but I tried to make the minimum set of changes to fix the issue. Happy to revisit though – it does seem odd to have the `__navigation` element not on the `<nav>`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -63,6 +63,19 @@ Do not include an empty `<span class="govuk-summary-list__actions"></span>` with
 
 This change was introduced in [pull request #2323: Avoid invalid nesting of `<span>` within a `<dd>` in summary list](https://github.com/alphagov/govuk-frontend/pull/2323).
 
+#### Update the HTML for navigation in the header
+
+We've updated the HTML for the header. This update only affects you if your header includes navigation.
+
+If you are not using Nunjucks macros then you should:
```
Comment (EoinShaughnessy) [inline CHANGELOG.md:70]:
```suggestion
If you're not using Nunjucks macros then you should:
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -63,6 +63,19 @@ Do not include an empty `<span class="govuk-summary-list__actions"></span>` with
...
+
+- move the `<button>` inside the `<nav>`, immediately before the `<ul>`
```
Comment (EoinShaughnessy) [inline CHANGELOG.md:72]:
```suggestion
- the `<button>` inside the `<nav>`, immediately before the `<ul>`
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -63,6 +63,19 @@ Do not include an empty `<span class="govuk-summary-list__actions"></span>` with
...
+- move the `aria-label` from the `<ul>` to the `<nav>`
```
Comment (EoinShaughnessy) [inline CHANGELOG.md:73]:
```suggestion
- the `aria-label` from the `<ul>` to the `<nav>`
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -63,6 +63,19 @@ Do not include an empty `<span class="govuk-summary-list__actions"></span>` with
...
+
+Check your changes against [the example in the Design System](https://design-system.service.gov.uk/components/header/#header-with-service-name-and-navigation).
+
+This change was introduced in [pull request #2427: Improvements to the screen reader experience for the menu in the header](https://github.com/alphagov/govuk-frontend/pull/2427).
```
Comment (EoinShaughnessy) [inline CHANGELOG.md:77]: Wonder could we shorten the PR title? Like so:

```suggestion
This change was introduced in [pull request #2427: Improvements to the screen reader experience for the header menu](https://github.com/alphagov/govuk-frontend/pull/2427).
```

**Review by EoinShaughnessy — COMMENTED** [review]
Update's looking good! Few small suggestions.

File: src/govuk/components/header/template.njk (hunk 1 above)
Comment (vanitabarrett) [reply src/govuk/components/header/template.njk:60]: Hm, tricky 🤔 Like you said, it seems a bit odd to be applying them to the `<ul>`. But I'm also not sure what people are using the `navigationClasses` for, which makes it difficult to know if they'd expect/find it useful for them to be applied to the `<ul>` or the `<nav>`. If we move it to the `<nav>` element, will we also need to provide a way for people to pass classes to the `<ul>`?

I'm not sure if we can separate the two - move the `govuk-header__navigation` class to the `<nav>` but keep the `navigationClasses` on the `<ul>`? Perhaps that's even more confusing though.

Comment (36degrees) [reply src/govuk/components/header/template.njk:60]: Yeah, I think it'd make sense to keep the two together.

Arguments for keeping `govuk-header__navigation` and `navigationClasses` on the `<ul>`:

- Preserves existing behaviour
- We don't know what people are using `navigationClasses` for, so it's hard to reason about what impact the change would have

Arguments for moving `govuk-header__navigation` and `navigationClasses` to the `<nav>`:

- Seems more 'correct' / what users would expect
- Users can still style the list if they need to by passing e.g. `app-header__navigation` as a `navigationClass` and then using `.app-header__navigation ul` to target it, but without this change would have no equivalent way to style the nav
- We could make it possible to add classes to the `<ul>` by introducing a new param, without having to make a breaking change
- If we didn't do this and later wanted to make it possible to style the `<nav>` we'd have to introduce a new param with a suboptimal name (e.g. `navigationWrapperClasses`) or come back and make this (breaking) change

Comment (vanitabarrett) [reply src/govuk/components/header/template.njk:60]: Thanks for summarising, @36degrees, that's really helpful! I don't think you've missed anything. It's a really good point about being able to style the `<ul>` by using the `navigationClasses` but not being able to do it the other way around - I think that's convinced me that we should make the change. I think we might need to provide some additional guidance or examples somewhere to help people make the change though, maybe in the Changelog

Comment (36degrees) [reply src/govuk/components/header/template.njk:60]: Yeah, I think I've talked myself into it too. Will push a commit…

**Comment by vanitabarrett** [comment]
This change looks really good! Think we just need to add something to the Changelog about the classes change?

**Comment by 36degrees** [comment]
> Think we just need to add something to the Changelog about the classes change?

I've added a section to the changelog entry which I think should cover it?

> Any additional classes passed using the `navigationClasses` Nunjucks option are now applied to the `<nav>` rather than the `<ul>`. Check that the additional classes are still doing what you expect.

**Review by vanitabarrett — APPROVED** [review]

**Comment by 36degrees** [comment]
@EoinShaughnessy are you happy with that latest addition to the release notes?

**Review by EoinShaughnessy — APPROVED** [review]
Looks good!

**Pull request merged**